python/ray/serve/_private/metrics_utils.py

```diff
@@ -307,71 +307,179 @@ def aggregate_avg(
+    for point in step_series:
+        if point.timestamp <= window_start:
+            continue  # Already processed above
+        if point.timestamp >= window_end:
+            break  # Beyond our window
+
+        # Add contribution of current segment
+        segment_end = min(point.timestamp, window_end)
+        duration = segment_end - current_time
+        if duration > 0:
+            total_weighted_value += current_value * duration
+            total_duration += duration
+
+        current_value = point.value
+        current_time = segment_end
+
+    # Add final segment if it extends to window_end
+    if current_time < window_end:
+        duration = window_end - current_time
+        total_weighted_value += current_value * duration
+        total_duration += duration
+
+    return total_weighted_value / total_duration if total_duration > 0 else None
+
+
+def merge_instantaneous_total(
+    replicas_timeseries: List[List[TimeStampedValue]],
+    ttl: Optional[float] = None,
 ) -> List[TimeStampedValue]:
     """
-    Merge two ascending time series by summing values within a specified time window.
-    If multiple values fall within the same window in a series, the latest value is used.
-    The output contains one point per window that had at least one value, timestamped
-    at the window center.
+    Merge multiple gauge time series (right-continuous, LOCF) into an
+    instantaneous total time series as a step function.
+
+    This approach treats each replica's gauge as right-continuous, last-observation-
+    carried-forward (LOCF), which matches gauge semantics. It produces an exact
+    instantaneous total across replicas without bias from arbitrary windowing.
+
+    Args:
+        replicas_timeseries: List of time series, one per replica. Each time series
+            is a list of TimeStampedValue objects sorted by timestamp.
+        ttl: Optional time-to-live for values. If specified, values expire after
+            this duration and are removed from the total.
+
+    Returns:
+        A list of TimeStampedValue representing the instantaneous total at event times.
+        Between events, the total remains constant (step function).
     """
-    if window_s <= 0:
-        raise ValueError(f"window_s must be positive, got {window_s}")
-
-    if not t1 and not t2:
+    if not replicas_timeseries or not any(replicas_timeseries):
         return []
 
-    # Align windows so each output timestamp sits at the start of its window.
-    # start is snapped to window_s boundary for binning stability
-    earliest = min(x[0].timestamp for x in (t1, t2) if x)
-    start = earliest // window_s * window_s
-
-    b1 = _bucket_latest_by_window(t1, start, window_s)
-    b2 = _bucket_latest_by_window(t2, start, window_s)
-
-    windows = sorted(set(b1.keys()) | set(b2.keys()))
-
+    # Build delta events per replica
+    # events[t] = total delta at time t across all replicas
+    events = defaultdict(float)  # time -> delta
```

Comment:
I'm a little worried about this because it's going to get huge - if the timestamps are being generated on different replicas there's almost no chance you'll have dupes if you don't round them, so this dict will almost surely have as many entries as there are entries in all the time series combined, and each entry will almost surely be populated by just one entry from one replica.

Reply:
good point, i think the amount of data we are generating if is function of (num_replicas, record_interval, push_interval, look_back_period). One thing we could do, in a follow up PR is to warn users if they choose bad values for their workload.

Also in this PR, I will add logs indicating the time spent in doing this calculation. That will tell us if there is a problem with performance.

Reply:
@abrarsheikh why not consider adding rounding?